transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,459 @@
+            if !me.send_incoming(muxer, cx) {
+                tracing::info!("cleared incoming slot");
+            }
+        }
+
+        if let Some((id, event)) = me.event_slot.take() {
+            if !me.send_event(id, event, cx) {
+                tracing::info!("cleared event slot");
+            }
+        }
+
+        while let Some(transmit) = me.endpoint.poll_transmit() {
+            me.transmit(transmit);
+        }
+
+        if me.event_slot.is_none() {
+            while let Poll::Ready(event) = me.channel.poll_next_event(cx) {
+                match event {
+                    Some(ToEndpoint::Dial {
+                        addr,
+                        public_key,
+                        tx,
+                    }) => {
+                        let crypto = C::new_client_config(&me.crypto_config, public_key);
+                        let client_config = ClientConfig {
+                            transport: me.crypto_config.transport.clone(),
+                            crypto,
+                        };
+                        let (id, connection) =
+                            match me.endpoint.connect(client_config, addr, "server_name") {
+                                Ok(c) => c,
+                                Err(err) => {
+                                    tracing::error!("dial failure: {}", err);
+                                    let _ = tx.send(Err(err.into()));
+                                    continue;
+                                }
+                            };
+                        let (channel, conn) = me.channel.create_connection(id);
+                        me.connections.insert(id, conn);
+                        let muxer = QuicMuxer::new(channel, connection);
+                        tx.send(Ok(muxer)).ok();
+                    }
+                    Some(ToEndpoint::ConnectionEvent {
+                        connection_id,
+                        event,
+                    }) => {
+                        let is_drained_event = event.is_drained();
+                        if is_drained_event {
+                            me.connections.remove(&connection_id);
+                        }
+                        if let Some(event) = me.endpoint.handle_event(connection_id, event) {
+                            if me.send_event(connection_id, event, cx) {
+                                tracing::info!("filled event slot");
+                                break;
+                            }
+                        }
+                    }
+                    Some(ToEndpoint::Transmit(transmit)) => {
+                        me.transmit(transmit);
+                    }
+                    None => {
+                        me.endpoint.reject_new_connections();
+                        return Poll::Ready(());
+                    }
+                }
+            }
+        }
+
+        while !me.outgoing.is_empty() {
+            me.outgoing.make_contiguous();
```

Comment:
This is O(n)

Reply:
`quinn` does the same. a queue built from two vecs will require popping and pushing a value once more than if we just used a stack using a single vec. not sure much can be done here, or that it is a problem in practice?

Reply:
No idea.  I think this is really a missing feature in the standard library, which should provide better APIs for this.

Reply:
since it's a queue, I don't think it can be done any more efficiently. as far as I know queues are usually built from two stacks, requiring pushing and popping the elements from one stack to the other. Make contiguous does just that.

Reply:
Marking this as resolved